Bookmark - Link Upsell flow

ashiagr opened the pull request:
## Description

This adds 

- Adds "Upsell view" on "Podcast details - Bookmarks tab"
- Links "Upsell view - subscribe" to upsell flow
- Adds "No bookmarks view" on "Podcast details - Bookmarks tab"
- Links "No bookmarks view" - Headset settings" button with the "Headphone controls" screen
- Adds appropriate guards for the bookmark feature

** There will be a few UI updates later related to Patron - early access for the feature. 

## Testing Instructions

Prerequisites: 

- Enable Patron + Bookmarks features in the Feature class.
- Release build
- Licence test account

#### Free/ Plus -  Upsell view

1. Install the app
2. Using free account, open the app and go to discover tab
3. Tap on a oodcast to go to `Podcast Details` view
4. ✅ Notice you see `Bookmarks` tab 
5. Tap on `Bookmarks` tab
6. ✅ Notice that you see `Upsell view`
7. Go to `Episode` tab and play an epsiode
8. Open full screen player
9. ✅ Notice that `Add bookmark` is not in shelf items
10. ✅ Notice you see `Bookmarks` tab with`Upsell view`
11. Dismiss the `Player` screen
12. Open the `Episode` screen
13. ✅ Notice you see `Bookmarks` tab with `Upsell view`
14. Go to `Profile` tab -> `Settings`
15. ✅ Notice there is no `Headset controls` setting

** `Headset controls` setting screen needs to be updated to show the paid feature icon for "Add bookmark" option

#### Patron - early access

1. Go to `Podcast details` screen -> `Bookmarks` tab
2. Tap on subscribe button
3. ✅ Notice you see Patron upsell flow
5. Subscribe to the plan using license test account
6. ✅ Once subscribed, notice that you return to `No bookmarks view`
7. Tap on `Headset settings` on `No bookmarks view`
8. ✅ Notice that you're taken to the `Headset controls` screen
9. Go back to `Podcast details` screen
10. From the Episodes tab, play an episode
11. Open the `Player` screen
12. ✅ Notice that there is an `Add bookmark` shelf item
13. Tap on `Add bookmark`
14. Go to `Bookmarks` tab 
15. ✅ Notice that you see bookmarks list
17. Return to `Podcast details` screen 
18. ✅ Notice that you see bookmark icon on the episode row
19. Logout
20. Go back to the episode row for the episode you added a bookmark to
21. ✅ Notice that you do not see the bookmark icon on the episode row


## Screenshots or Screencast 

Captured in pdeCcb-25g-p2#comment-2434

## Checklist
- [ ] If this is a user-facing change, I have added an entry in CHANGELOG.md
- [ ] Ensure the linter passes (`./gradlew spotlessApply` to automatically apply formatting/linting)
- [ ] I have considered whether it makes sense to add tests for my changes
- [ ] All strings that need to be localized are in `modules/services/localization/src/main/res/values/strings.xml`
- [ ] Any jetpack compose components I added or changed are covered by compose previews
 
#### I have tested any UI changes...
<!-- If this PR does not contain UI changes, ignore these items -->
- [ ] with different themes
- [ ] with a landscape orientation
- [ ] with the device set to have a large display and font size
- [ ] for accessibility with TalkBack

mchowning reviewed (approved):
This is looking good. Bookmarks are going to be great! 😄 

Here's a brain dump of some things that I noticed while testing this. Some of these are not directly related to the changes in this PR though, and I think that any of them that need t can be addressed in separate PRs, so I'll go ahead and approve.

### 1. Headphone controls and free accounts
Under the test steps for free/plus accounts we have:
> - Go to Profile tab -> Settings
> - ✅ Notice there is no Headset controls setting
I'm still seeing the headphone controls option, and I can update one of the headphone control options is "Add bookmark".

### 2. Bookmarks tab on discover screen
Are these test steps correct?
> - Using free account, open the app and go to discover tab
> - ✅ Notice you see Bookmarks tab
I didn't see a bookmarks tab on the discover screen, but I'm guessing this was probably just a typo in the testing steps.

### 3. Patron bookmark upsell
The Patron upsell message on the bookmarks screen looks like it is basically the same as what we were going to use for the Plus upsell once bookmarks makes it to Plus. That concerns me a bit because it makes it seem like bookmarks is a Patron-only feature as opposed to a Patron-only-for-now feature. I think it would be good to add something to this upsell message about subscribing to Patron to get _early access_ to bookmarks.

![image](https://github.com/Automattic/pocket-casts-android/assets/4656348/d5ae08e2-cb2b-4aaf-b76d-edc73fb205ff)

### 3. Refresh failures

I've noticed some fairly frequent refresh failures when interacting with Patron/bookmarks. For some reason, I was only seeing it on my release builds now, but previously when I saw them they seemed to be related to bookmark syncing (sorry I don't have a stacktrace anymore since it's only happening for me on release builds now). 

If you're not experiencing these at all, then maybe it's just something weird that is going on with my build setup.

### 4. No bookmarks view text
We use the same "No bookmarks yet" view for the podcast screen, the episode screen, and the now playing screen. Since the episode screen and the now playing screen will be empty if you don't have any bookmarks for that episode (but you may have bookmarks for different episodes) it feels to me like it would be good to tweak the "No bookmarks" text on these screens to be something more specific like "No bookmarks for this episode" and "No bookmarks for this podcast".

The current "No bookmarks yet" text would make perfect sense to me if/when we have a screen that would display all of a user's bookmarks across all their podcasts.

### 5. Bookmarks icon on episode list stays after bookmark removed if the episode is paused

I think this is a pretty minor edge case where you have to follow a fairly specific flow, but when deleting a bookmark from the episode details screen, it seems that the bookmark icon on the episode list screen doesn't get removed if playback is paused. 

This is very much an edge case: I do not observe this issue if I delete the bookmark from a different screen or if the episode is actively playing (as soon as I tap play or do something else that causes the episode list to refresh, the bookmark icon gets properly removed from the episode list screen).

<details><summary>Video</summary>

https://github.com/Automattic/pocket-casts-android/assets/4656348/bce0271a-e5d6-4383-bc0d-7ac76511143f

</details>

### 6. Can't access bookmarks when signed out

After adding bookmarks, if I sign out I can no longer access them. I thought that earlier when I tested this I could access them when signing back in even if the account had become a free account, but testing this just now, I only get access to the bookmarks once my account is Patron again. Is that intentional? I think it would be nice if users could still access their bookmarks with a free account, they just wouldn't be able to add new ones.

<details><summary>Somewhat meandering video of me signing out and signing back in after my Patron subscription expired (UI shows I'm still Patron at the beginning probably because of the refresh issues that I've been seeing)</summary>



https://github.com/Automattic/pocket-casts-android/assets/4656348/a5828eb4-dfec-47ef-9111-39d111decd31



</details>

### 7. Upgrade prompt to google sign in exits onboarding flow

Tapping on bookmarks upgrade prompt when not signed in, I am taken to sign in, and then once I sign into a free account using Google sign in, I am returned to the bookmarks screen instead of taken on through the onboarding flow. This sounds similar to this issue: https://github.com/Automattic/pocket-casts-android/issues/1184. Maybe there will be a similar fix.

Also, I don't think this is a bookmarks-specific bug. I can see similar behavior with the folder upgrade prompt if I use Google sign in.

mchowning commented:
> ### 7. Upgrade prompt to google sign in exits onboarding flow
> 
> Tapping on bookmarks upgrade prompt when not signed in, I am taken to sign in, and then once I sign into a free account using Google sign in, I am returned to the bookmarks screen instead of taken on through the onboarding flow. This sounds similar to this issue: https://github.com/Automattic/pocket-casts-android/issues/1184. Maybe there will be a similar fix.
> 
> Also, I don't think this is a bookmarks-specific bug. I can see similar behavior with the folder upgrade prompt if I use Google sign in.

I think I may have this fixed https://github.com/Automattic/pocket-casts-android/pull/1275, but I need to do some more testing.

ashiagr commented:
Thanks for the thorough testing and feedback, @mchowning! 🙏  I appreciate your time. 

### 1. Headphone controls and free accounts

> I'm still seeing the headphone controls option, and I can update one of the headphone control options is "Add bookmark".

Actually, it should be present but should not allow selecting `Add bookmark` for non-Patron users. I added it as a note in [this project task](https://github.com/orgs/Automattic/projects/692/views/3?pane=issue&itemId=35684105), and I'll address it as part of the task where I'll make "Add bookmark" a paid option under "Headphone controls" setting.

### 2. Bookmarks tab on discover screen
> Are these test steps correct?

I updated testing steps to open the `Podcast Details` page from the `Discover` tab to see the tabs. I missed that step. 🙈

### 3. Patron bookmark upsell
> The Patron upsell message on the bookmarks screen looks like it is basically the same as what we were going to use for the Plus upsell once bookmarks makes it to Plus. That concerns me a bit because it makes it seem like bookmarks is a Patron-only feature as opposed to a Patron-only-for-now feature. I think it would be good to add something to this upsell message about subscribing to Patron to get _early access_ to bookmarks.

You're right. I mentioned in the PR description that we need to make some UI adjustments to make it clear that it is an early-access feature for `Patron` users. I'll address it in a separate PR. (Also started a Slack discussion: p1692162285498199-slack-C055BDMU095)

### 3. Refresh failures

I added a [task](https://github.com/orgs/Automattic/projects/692/views/3?pane=issue&itemId=36022880) to test and investigate refresh failures. 

### 4. No bookmarks view text

I added a [task](https://github.com/orgs/Automattic/projects/692/views/3?pane=issue&itemId=36023042) to update the text after coordinating with iOS devs.

### 5. Bookmarks icon on episode list stays after bookmark removed if the episode is paused

> This does seem somewhat similar to https://github.com/Automattic/pocket-casts-android/issues/1273.

We're looking into it in #1273. Thanks for pointing it out.

### 6. Can't access bookmarks when signed out

> After adding bookmarks, if I sign out I can no longer access them.

Discussing it here (p1692162947882119-slack-C055BDMU095), so far it is an intended behavior and matches the behavior on iOS.

### 7. Upgrade prompt to google sign in exits onboarding flow

> I think I may have this fixed here (https://github.com/Automattic/pocket-casts-android/pull/1275), but I need to do some more testing.

Yes, you fixed it already! 🎉  I merged `main` branch and tested the scenario, it works for google sign-in.

<details><summary>Video</summary>

https://github.com/Automattic/pocket-casts-android/assets/1405144/3dd34243-52d4-49f5-b1a9-cbad5e5bb1ac

</details>

ashiagr commented on modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/bookmark/BookmarkHelper.kt:
```diff
@@ -67,6 +70,12 @@ class BookmarkHelper @Inject constructor(
+
+    private fun shouldAllowAddBookmark(): Boolean {
+        return settings.getCachedSubscription()?.let { subscriptionStatus ->
+            subscriptionStatus is SubscriptionStatus.Paid && subscriptionStatus.tier == SubscriptionTier.PATRON
```
Oh, it was just to type cast so that I could access the tier. I changed it slightly in https://github.com/Automattic/pocket-casts-android/pull/1274/commits/84701c44636a882781021e37df110845b6a57099 so that it is more clear.

ashiagr commented on modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/view/bookmark/BookmarksPage.kt:
```diff
@@ -107,8 +117,20 @@ private fun Content(
+                    .fillMaxHeight()
+                    .verticalScroll(rememberScrollState())
+            )
+            is UiState.PlusUpsell -> PlusUpsellView(
```
Dropped in https://github.com/Automattic/pocket-casts-android/pull/1274/commits/e3f3e1a33b49760121c54e35075654f6d0d1a757 and https://github.com/Automattic/pocket-casts-android/pull/1274/commits/bb71674ba321248b90b3621cd8628a51516ada00

ashiagr commented on modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/view/bookmark/BookmarksFragment.kt:
```diff
@@ -172,4 +189,10 @@ class BookmarksFragment : BaseFragment() {
+
+    private fun onUpgradeClicked() {
+        val source = OnboardingUpgradeSource.BOOKMARKS
+        val onboardingFlow = OnboardingFlow.PlusUpsell(source, true)
```
Renamed in https://github.com/Automattic/pocket-casts-android/pull/1274/commits/dc2dc2c9da07b6579ff6849f876cd804a7069ddd
